File: packages/sdk/src/numbers.mjs
// Numeric admission for values that arrived from somewhere else.
//
// WHY THIS FILE EXISTS, AND WHY IT IS NOT `??`.
//
// `x ?? fallback` defends `undefined` and `null` and NOTHING ELSE. Every other wrong
// shape sails straight through it:
//
//     const waitMs = body.waitMs ?? 2000;        // "5000"  -> a string
//     await sleep(waitMs);                       // NaN     -> setTimeout fires immediately
//     if (waitMs > LIMIT) ...                    // NaN > n -> false, so the cap fails OPEN
//
// Every one of those is a value a server, a JSON file, a CLI flag or a plugin can hand
// this SDK, and each turns a guard into decoration. `>` against a non-finite number is
// always false, so the comparison that was supposed to be the limit stops being one and
// reports nothing while it happens.
//
// So: numbers that cross a trust boundary come through here, and anything that is not a
// real finite number of the right shape is REPLACED by the caller's fallback rather than
// used. The fallback is a parameter and never a silent zero, because "we could not read
// it" and "it was zero" are different facts and only one of them is safe to sleep on.

/** True only for a real, finite JS number. Rejects NaN, ±Infinity, strings, null, bigint. */
export function isFiniteNumber(value) {
  return typeof value === 'number' && Number.isFinite(value);
}

/**
 * A finite number inside [min, max], or `fallback`.
 *
 * Out-of-range finite values are CLAMPED, not rejected: a server asking for a 10-minute
 * poll delay wants a long delay, and answering it with the default would poll 300x more
 * often than it asked. A value that is not a number at all is a different case and gets
 * the fallback, because nothing about it is meaningful.
 */
export function finiteNumber(value, fallback, { min = -Infinity, max = Infinity } = {}) {
  if (!isFiniteNumber(value)) return fallback;
  if (value < min) return min;
  if (value > max) return max;
  return value;
}

/** As `finiteNumber`, but the result is an integer (truncated toward zero). */
export function finiteInt(value, fallback, range = {}) {
  const n = finiteNumber(value, fallback, range);
  return Math.trunc(n);
}

/**
 * Seconds from a `Retry-After` header, or null.
 *
 * The header is legally either a delta in seconds or an HTTP-date, and both forms show up
 * in the wild. A date in the past yields 0 — meaning "retry now" — never a negative delay
 * that a later `Math.min` would happily pick as the smallest.
 */
export function retryAfterSeconds(raw) {
  if (typeof raw !== 'string' || raw.trim() === '') return null;
  const asNumber = Number(raw.trim());
  if (Number.isFinite(asNumber)) return Math.max(0, asNumber);
  const asDate = Date.parse(raw);
  if (Number.isNaN(asDate)) return null;
  return Math.max(0, (asDate - Date.now()) / 1000);
}

File: packages/sdk/python/golem_sdk/numbers.py
"""Numeric admission for values that arrived from somewhere else.

The Python twin of ``src/numbers.mjs``, and it exists for the same reason. Python's
``or`` is the local equivalent of JavaScript's ``??`` and is worse: ``data.get("waitMs")
or 2000`` replaces a *legitimate zero* as well as a missing value, and ``float("nan")``
is truthy, so it sails through and then loses every comparison it takes part in::

    wait = data.get("waitMs") or 2000     # NaN survives
    if wait > CEILING:                    # NaN > n is False -> the cap fails open
        wait = CEILING
    time.sleep(wait)                      # ValueError, from a line that never mentions wait

A value that is not a real finite number is REPLACED by the caller's fallback, never used.
"""
from __future__ import annotations

import math
from typing import Any, Optional

# bool is a subclass of int in Python, and True would otherwise pass as the number 1.
_NUMERIC = (int, float)


def is_finite_number(value: Any) -> bool:
    """True only for a real, finite int or float. Rejects NaN, inf, bools and strings."""
    if isinstance(value, bool) or not isinstance(value, _NUMERIC):
        return False
    return math.isfinite(value)


def finite_number(
    value: Any,
    fallback: float,
    minimum: float = -math.inf,
    maximum: float = math.inf,
) -> float:
    """A finite number inside ``[minimum, maximum]``, or ``fallback``.

    A real number out of range is CLAMPED rather than discarded: a server asking for a
    ten-minute delay means it, and answering with the default would poll far more often
    than it asked. Something that is not a number at all gets the fallback, because
    nothing about it is meaningful.
    """
    if not is_finite_number(value):
        return fallback
    return max(minimum, min(maximum, float(value)))


def finite_int(value: Any, fallback: int, minimum: float = -math.inf, maximum: float = math.inf) -> int:
    return int(finite_number(value, fallback, minimum, maximum))


def retry_after_seconds(raw: Optional[str]) -> Optional[float]:
    """Seconds from a ``Retry-After`` header, in either legal form, or None.

    A date already in the past yields 0 — "retry now" — never a negative delay that a
    later ``min()`` would happily choose.
    """
    if not isinstance(raw, str) or raw.strip() == "":
        return None
    try:
        return max(0.0, float(raw.strip()))
    except ValueError:
        pass
    from email.utils import parsedate_to_datetime
    import datetime

    try:
        when = parsedate_to_datetime(raw)
    except (TypeError, ValueError):
        return None
    if when is None:
        return None
    if when.tzinfo is None:
        when = when.replace(tzinfo=datetime.timezone.utc)
    return max(0.0, (when - datetime.datetime.now(datetime.timezone.utc)).total_seconds())
